java/calcite/src/main/java/org/apache/calcite/rel/externalize/MapDRelJson.java

```diff
@@ -458,7 +465,33 @@ RexNode toRex(RelInput relInput, Object o) {
         if (literal == null) {
           return rexBuilder.makeNullLiteral(sqlTypeName);
         }
-        return toRex(relInput, literal);
+
+        // omnisci serializes numeric literals differently, we need more data
+        // then the number. So, we need to have a special case for that.
+        if (literal instanceof Number) {
+          final SqlTypeName targetTypeName =
+                  Util.enumVal(SqlTypeName.class, (String) map.get("target_type"));
+          final long scale = ((Number) map.get("scale")).longValue();
+          final long precision = ((Number) map.get("precision")).longValue();
+          final long typeScale = ((Number) map.get("type_scale")).longValue();
+          final long typePrecision = ((Number) map.get("type_precision")).longValue();
+          RelDataTypeFactory typeFactory = cluster.getTypeFactory();
+
+          BigDecimal value =
+                  BigDecimal.valueOf(((Number) literal).longValue(), (int) scale);
+
+          if (typeScale != 0 && typeScale != -2147483648) {
```

Comment:
Surprised to see the null sentinel here in the `type_scale` field. Can you explain?

Reply:
As we found out, omnisci format has type_scale field even if literal is not decimal. At that case typeScale field would have the value of -2147483648.
So, we have the same logic in our calcite serializer [here](https://github.com/modin-project/modin/blob/623de5c3c90f9d44262f5ce4e9d7bae6d54e5dc8/modin/experimental/engines/omnisci_on_ray/frame/calcite_serializer.py#L132).